### PR TITLE
[Serve] Incorporate `target_capacity` changes into Serve's statuses

### PR DESCRIPTION
**Pull request by shrekris-anyscale**
## Why are these changes needed?

This change makes Serve incorporate `target_capacity` changes when calculating statuses. E.g., if `target_capacity` changes, then Serve deployments enter the `UPDATING` state until their running number of replicas adheres to the new scaled `num_replicas` value.

## Related issue number

N/A

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
     - This change adds unit tests and updates existing tests.

### CODE REVIEW

**Comment by edoakes** [comment]
Ready for review @shrekris-anyscale?

File: python/ray/serve/_private/autoscaling_policy.py [hunk 1]
```diff
@@ -202,3 +221,46 @@ def get_decision_num_replicas(
...
+        curr_lower_bound = self.get_curr_lower_bound()
+
+        return max(curr_lower_bound, min(curr_upper_bound, curr_target_num_replicas))
+
+    def get_curr_lower_bound(self) -> int:
+        if self.get_capacity_adjusted_initial_replicas() is not None and (
+            self.scale_direction is None
+            or self.scale_direction == TargetCapacityScaleDirection.UP
+        ):
+            return self.get_capacity_adjusted_initial_replicas()
```
Comment (zcin) [inline python/ray/serve/_private/autoscaling_policy.py:264]: Why does initial replicas factor into the lower bound during autoscaling? Shouldn't this just be set when the deployment is deployed/updated?

File: python/ray/serve/_private/autoscaling_policy.py [hunk 2]
```diff
@@ -1,16 +1,20 @@
...
 from ray.serve.config import AutoscalingConfig
 
 logger = logging.getLogger(SERVE_LOGGER_NAME)
 
 
 def calculate_desired_num_replicas(
-    autoscaling_config: AutoscalingConfig, current_num_ongoing_requests: List[float]
+    autoscaling_config: AutoscalingConfig,
+    current_num_ongoing_requests: List[float],
+    override_min_replicas: Optional[float] = None,
```
Comment (zcin) [inline python/ray/serve/_private/autoscaling_policy.py:17]: Should there also be an `override_max_replicas`?

**Review by zcin — COMMENTED** [review]
Great work!

File: python/ray/serve/_private/autoscaling_policy.py (hunk 2 above)
Comment (shrekris-anyscale) [reply python/ray/serve/_private/autoscaling_policy.py:17]: Thanks for the catch, I added it.

File: python/ray/serve/_private/autoscaling_policy.py (hunk 1 above)
Comment (shrekris-anyscale) [reply python/ray/serve/_private/autoscaling_policy.py]: While the `target_capacity` is scaling up, we scale `initial_replicas` by `target_capacity` and use that as a lower bound. This makes sense semantically because `initial_replicas` should be the number of replicas that a deployment starts with, and a deployment hasn't quite started until `target_capacity` has finished scaling up.

File: python/ray/serve/_private/deployment_state.py [hunk 3]
```diff
@@ -89,40 +89,35 @@ class DeploymentTargetState:
     """The current goal state for a deployment.
 
     info: contains the information needed to initialize a replica.
-    num_replicas: the number of replicas to run.
-    adjust_capacity: whether the number of replicas should be adjusted by
-        the current target_capacity.
+    target_num_replicas: the number of replicas to run. This should already
```
[the remark this answers is not in the document]
Comment (shrekris-anyscale) [reply python/ray/serve/_private/deployment_state.py:92]: I don't want to overload the term `num_replicas`. Now, `num_replicas` is the value that the user sets, and `target_num_replicas` is the number of replicas that the deployment state manager is attempting to run at a particular moment.

File: python/ray/serve/_private/deployment_state.py [hunk 4]
```diff
@@ -1498,56 +1468,74 @@ def deploy(
...
-            # loop.
-            adjust_capacity = True
+            logger.info(
+                f"not exiting early: {target_capacity} "
+                f"{target_capacity_scale_direction} "
+            )
+
+        deployment_info.set_target_capacity(
+            target_capacity, target_capacity_scale_direction
+        )
```
[the remark this answers is not in the document]
Comment (shrekris-anyscale) [reply python/ray/serve/_private/deployment_state.py:1503]: We need to pass in the `target_capacity` and `target_capacity_scale_direction` into the `deploy()` method anyways, so we can compare against the deployment's existing `target_capacity` and `scale_direction` and decide whether to exit early or not ([code source](https://github.com/shrekris-anyscale/ray/blob/7f1b8e91d2e3c1a831648b86812978cc7bec565a/python/ray/serve/_private/deployment_state.py#L1495-L1497)). I'd prefer to also set the target capacity in the deployment state to keep everything in one place.

[the remark this answers is not in the document]
Comment (zcin) [reply python/ray/serve/_private/deployment_state.py:1503]: Could you instead do the following comparison to avoid having to pass them into the deploy method?
```
                and curr_deployment_info.target_capacity == deployment_info.target_capacity
                and curr_deployment_info.target_capacity_scale_direction
                == deployment_info.target_capacity_scale_direction
```

File: python/ray/serve/_private/deployment_state.py [hunk 5]
```diff
@@ -2227,16 +2175,14 @@ def update(
             # we manage.
 
             # Check the state of existing replicas and transition if necessary.
-            self._check_and_update_replicas(adjusted_target_num_replicas)
+            self._check_and_update_replicas(target_num_replicas)
```
[the remark this answers is not in the document]
Comment (shrekris-anyscale) [reply python/ray/serve/_private/deployment_state.py]: Good catch, I made that call and the ones after it get the target state directly instead of passing it in.

File: python/ray/serve/_private/deployment_state.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (shrekris-anyscale) [reply python/ray/serve/_private/deployment_state.py]: Nice, yeah that's a great idea. I set the `target_capacity` and `target_capacity_scale_direction` in the application state manager.

File: python/ray/serve/_private/application_state.py [hunk 6]
```diff
@@ -69,6 +69,9 @@ class BuildAppTaskInfo:
 
     obj_ref: ObjectRef
     code_version: str
+    config: ServeApplicationSchema
+    target_capacity: Optional[float]
+    target_capacity_scale_direction: Optional[TargetCapacityScaleDirection]
```
[the remark this answers is not in the document]
Comment (shrekris-anyscale) [reply python/ray/serve/_private/application_state.py]: I renamed `target_capacity_scale_direction` to `target_capacity_direction`.

File: python/ray/serve/_private/utils.py [hunk 7]
```diff
@@ -619,3 +620,27 @@ def time(self) -> float:
 class Timer(TimerBase):
     def time(self) -> float:
         return time.time()
+
+
+def get_capacity_adjusted_num_replicas(
```
Comment (edoakes) [inline python/ray/serve/_private/utils.py:625]: This feels like it does not belong in `utils.py`, but rather `deployment_state.py`

File: python/ray/serve/tests/unit/test_deployment_state.py [hunk 8]
```diff
@@ -3463,67 +3545,105 @@ def test_reduce_target_capacity(self, mock_deployment_state):
...
+            target_capacity=0,
+            target_capacity_direction=TargetCapacityDirection.UP,
+        )
+
+        deployment_state.update()
         check_counts(deployment_state, total=0)
         assert deployment_state.curr_status_info.status == DeploymentStatus.HEALTHY
 
-        # Increase target_capacity to 0, should have 1 replica start up.
-        deployment_state_update_result = deployment_state.update(target_capacity=1)
```
Comment (edoakes) [inline python/ray/serve/tests/unit/test_deployment_state.py:3509]: oops, good catch :)

File: python/ray/serve/tests/test_target_capacity.py [hunk 9]
```diff
@@ -373,6 +380,358 @@ def test_autoscaling_scale_to_zero(
...
+    async def __del__(self):
+        # Use try-except here to avoid cluttering logs if the SignalActor
+        # gets torn down before the replicas during clean up.
+        try:
+            await self.lifecycle_signal.wait.remote()
+        except RayActorError:
+            print("Lifecycle signal actor already dead. Replica exiting.")
+
+
+def create_controlled_app(config: Dict) -> Application:
```
Comment (edoakes) [inline python/ray/serve/tests/test_target_capacity.py:411]: Better to use pydantic-typed `config` here :)

File: python/ray/serve/tests/test_target_capacity.py [hunk 10]
```diff
@@ -373,6 +380,358 @@ def test_autoscaling_scale_to_zero(
...
+        client: ServeControllerClient,
+        target_capacity: Optional[float],
+        config: ServeDeploySchema,
+        app_name: str,
+        deployment_name: str,
+        check_statuses: bool = True,
+    ):
+        """Applies config with specified target_capacity."""
+
+        config.target_capacity = target_capacity
```
Comment (edoakes) [inline python/ray/serve/tests/test_target_capacity.py:459]: make a copy to avoid mutating passed-in arg?

File: python/ray/serve/tests/test_target_capacity.py [hunk 11]
```diff
@@ -373,6 +380,358 @@ def test_autoscaling_scale_to_zero(
...
+                lambda: serve.status()
+                .applications[app_name]
+                .deployments[deployment_name]
+                .status
+                in [
+                    DeploymentStatus.UPDATING,
+                    DeploymentStatus.UPSCALING,
+                    DeploymentStatus.DOWNSCALING,
+                ]
+            )
```
Comment (edoakes) [inline python/ray/serve/tests/test_target_capacity.py:477]: these conditions aren't really that helpful given they're so permissive (allowing for any of the possible status). can we instead pass in the expected one to this method?

File: python/ray/serve/tests/test_target_capacity.py [hunk 12]
```diff
@@ -373,6 +380,358 @@ def test_autoscaling_scale_to_zero(
...
+                DeploymentStatusTrigger.AUTOSCALING,
+            ]
+
+    def start_replicas_and_check_status(self, signal, app_name: str):
+        ray.get(signal.send.remote())
+        wait_for_condition(
+            lambda: serve.status().applications[app_name].status
+            == ApplicationStatus.RUNNING
+        )
+        ray.get(signal.send.remote(clear=True))
```
Comment (edoakes) [inline python/ray/serve/tests/test_target_capacity.py:491]: this method is pretty much unintelligible on its own. I don't know why sending the signal is expected to cause the application to enter running / start replicas. perhaps better naming is in order

File: python/ray/serve/tests/test_target_capacity.py [hunk 13]
```diff
@@ -373,6 +380,358 @@ def test_autoscaling_scale_to_zero(
...
+        app_name = "controlled_app"
+        deployment_name = "controlled"
+        num_replicas = 20
+
+        signal = SignalActor.options(
+            name="lifecycle_signal", namespace=SERVE_NAMESPACE
+        ).remote()
+        _ = SignalActor.options(
+            name="request_signal", namespace=SERVE_NAMESPACE
+        ).remote()
```
Comment (edoakes) [inline python/ray/serve/tests/test_target_capacity.py:507]: why do we need to start this if we don't use it?

Comment (edoakes) [reply python/ray/serve/tests/test_target_capacity.py:507]: if it's because the deployment fails without it existing, you can have that just catch the `ValueError`

File: python/ray/serve/tests/test_target_capacity.py [hunk 14]
```diff
@@ -373,6 +380,358 @@ def test_autoscaling_scale_to_zero(
...
+            metrics_interval_s=0.01,
+            look_back_period_s=0.01,
+            upscale_delay_s=0.01,
+            downscale_delay_s=0.01,
+        ),
+        graceful_shutdown_timeout_s=0,
+    ).bind()
+
+
+class TestTargetCapacityUpdateAndServeStatus:
```
Comment (edoakes) [inline python/ray/serve/tests/test_target_capacity.py:439]: These test cases look very redundant with the tests written above. Is there a reason not to just update those tests to check the applicable status instead?

File: python/ray/serve/_private/deployment_info.py [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+    TargetCapacityDirection.UP: TargetCapacityDirectionProtoEnum.UP,
+    TargetCapacityDirection.DOWN: TargetCapacityDirectionProtoEnum.DOWN,
+}
+
+
+TARGET_CAPACITY_DIRECTION_FROM_PROTO = {
+    TargetCapacityDirectionProtoEnum.UNSET: None,
+    TargetCapacityDirectionProtoEnum.UP: TargetCapacityDirection.UP,
+    TargetCapacityDirectionProtoEnum.DOWN: TargetCapacityDirection.DOWN,
+}
```
Comment (edoakes) [inline python/ray/serve/_private/deployment_info.py:31]: Why do we need to have both enums? Can we instead just use the proto one?

File: python/ray/serve/_private/deployment_info.py [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+
+
+class DeploymentInfo:
```
Comment (edoakes) [inline python/ray/serve/_private/deployment_info.py:34]: I assume this is basically copy-pasted from `common.py` -- is there a reason it needs to be done in this PR?

File: python/ray/serve/_private/application_state.py [hunk 17]
```diff
@@ -353,33 +358,32 @@ def deploy_config(
...
                 self._update_status(
                     ApplicationStatus.DEPLOY_FAILED, traceback.format_exc()
                 )
             except Exception:
-                self._set_target_state(
-                    deployment_infos=None,
-                    code_version=None,
-                    target_config=self._target_state.config,
-                )
+                self._clear_target_state()
```
Comment (edoakes) [inline python/ray/serve/_private/application_state.py:386]: This appears to be a behavior change -- previously we set `target_config=self._target_state.config` and now it's cleared to `None`. What's the purpose for this change and are you sure it doesn't have side effects?

File: python/ray/serve/_private/autoscaling_policy.py [hunk 18]
```diff
@@ -148,6 +163,13 @@ def __init__(self, config: AutoscalingConfig):
         # scale_up_periods or scale_down_periods.
         self.decision_counter = 0
 
+        # Scales the min_replicas and max_replicas when calculating the final
+        # number of replicas.
+        self.target_capacity: Optional[float] = None
+
+        # Determines what lower bound to use when calculating num_replicas
```
Comment (edoakes) [inline python/ray/serve/_private/autoscaling_policy.py:170]: expand the comment to indicate what lower bound is used for each direction

File: python/ray/serve/_private/autoscaling_policy.py [hunk 19]
```diff
@@ -202,3 +227,54 @@ def get_decision_num_replicas(
...
 
         return decision_num_replicas
+
+    def set_target_capacity(
+        self,
+        new_target_capacity: Optional[float],
+        new_target_capacity_direction: Optional[TargetCapacityDirection],
+    ):
+        self.target_capacity = new_target_capacity
+        self.target_capacity_direction = new_target_capacity_direction
```
Comment (edoakes) [inline python/ray/serve/_private/autoscaling_policy.py:237]: is there a reason this needs to be made stateful rather than passing it into the `calculate_desired_num_replicas` call?

File: python/ray/serve/_private/autoscaling_policy.py [hunk 20]
```diff
@@ -202,3 +227,54 @@ def get_decision_num_replicas(
...
+            return get_capacity_adjusted_num_replicas(
+                self.config.initial_replicas, self.target_capacity
+            )
+
+    def get_capacity_adjusted_max_replicas(self) -> int:
+        return get_capacity_adjusted_num_replicas(
+            self.config.max_replicas, self.target_capacity
+        )
+
+    def apply_initial_bounds(self, curr_target_num_replicas: int) -> int:
```
Comment (edoakes) [inline python/ray/serve/_private/autoscaling_policy.py:257]: it's unclear to me what "initial" means here

File: python/ray/serve/_private/deployment_state.py [hunk 21]
```diff
@@ -1465,88 +1459,59 @@ def _set_target_state(
...
-                if existing_info is not None:
-                    autoscaled_num_replicas = self._target_state.num_replicas
-                else:
-                    autoscaled_num_replicas = autoscaling_config.min_replicas
-            autoscaled_num_replicas = self.get_capacity_adjusted_num_replicas(
-                autoscaled_num_replicas, target_capacity
-            )
-            deployment_info.set_autoscaled_num_replicas(autoscaled_num_replicas)
+                target_num_replicas = autoscaling_policy.apply_initial_bounds(
+                    self._target_state.target_num_replicas
```
Comment (edoakes) [inline python/ray/serve/_private/deployment_state.py:1503]: I'm struggling to understand how this interacts with `get_decision_num_replicas` -- why are both needed and the calls made in different places?

File: python/ray/serve/_private/deployment_info.py (hunk 16 above)
Comment (shrekris-anyscale) [reply python/ray/serve/_private/deployment_info.py:34]: Yes, it's copy-pasted. There was a circular import otherwise. Without the copy-paste:

* `common.py` defined:
    * `DeploymentInfo`, which depends on
        * `BasicAutoscalingPolicy`, which is stored in `autoscaling_policy.py`.
    * `TargetCapacityDirection`
* `autoscaling_policy.py`, which depends on
    * `TargetCapacityDirection`, which is defined in `common.py`.

When `DeploymentInfo` is in `common.py`, `common.py` needs to import `autoscaling_policy.py`. But then `autoscaling_policy.py` needs to import `common.py`.

File: python/ray/serve/_private/application_state.py (hunk 17 above)
Comment (shrekris-anyscale) [reply python/ray/serve/_private/application_state.py:386]: Good catch, I didn't see a reason that we were setting `target_config` to `self._target_state.config`, and the reason for these calls seemed to be to clear the target state altogether.

I tried reading through the code, but I didn't see any reason to keep the `self._target_state.config`, and all unit tests continue to pass even when I remove it.

@zcin could you confirm whether this behavior change is okay?

File: python/ray/serve/_private/autoscaling_policy.py (hunk 20 above)
Comment (shrekris-anyscale) [reply python/ray/serve/_private/autoscaling_policy.py]: This was meant to apply the bounds when a deployment was originally deployed. I've refactored the logic a bit (see [`deployment_state.py`](https://github.com/ray-project/ray/pull/41317/files#diff-c4c2583a4c2f3a3c87ada6faebd0b2dfef404e165df133a11195be5d65fcb387R1472-R1525) and [`autoscaling_policy.py`](https://github.com/ray-project/ray/pull/41317/files#diff-02afedfa9ea08814a5f985924237f709cb5eea88d611d4687f34dd617dddcc01R260-R279)), so hopefully it's easier to follow. Let me know what you think.

File: python/ray/serve/_private/deployment_state.py (hunk 21 above)
Comment (shrekris-anyscale) [reply python/ray/serve/_private/deployment_state.py]: I've refactored this logic a bit. See [this reply](https://github.com/ray-project/ray/pull/41317#discussion_r1409841115) for links.

File: python/ray/serve/_private/autoscaling_policy.py (hunk 19 above)
Comment (edoakes) [reply python/ray/serve/_private/autoscaling_policy.py:240]: I see, it's also used in `get_current_lower_bound`

Comment (shrekris-anyscale) [reply python/ray/serve/_private/autoscaling_policy.py:240]: Yes, but all these methods can be made stateless. I'm working on that right now.

Comment (shrekris-anyscale) [reply python/ray/serve/_private/autoscaling_policy.py:248]: I made it stateless.

File: python/ray/serve/_private/autoscaling_policy.py (hunk 20 above)
Comment (edoakes) [reply python/ray/serve/_private/autoscaling_policy.py]: Much better!

**Comment by edoakes** [comment]
Core logic LGTM, will approve once comments on the tests are addressed and Cindy confirms there are no unintended effects of changing the `clear_target_state` behavio.

**Comment by shrekris-anyscale** [comment]
Sounds good, I'll work on addressing the test suggestions.

File: python/ray/serve/_private/deployment_info.py (hunk 15 above)
Comment (shrekris-anyscale) [reply python/ray/serve/_private/deployment_info.py]: I simplified the conversion. I pattern-matched `DeploymentStatus`, which also uses a separate enum for the protobuf and the Python code.

File: python/ray/serve/_private/utils.py (hunk 7 above)
Comment (shrekris-anyscale) [reply python/ray/serve/_private/utils.py:625]: This method is used in both `deployment_state.py` and `autoscaling_policy.py`. If we move it to `deployment_state.py`, then `autoscaling_policy.py` would have to import `deployment_state.py`, which seems like an awkward dependency structure. I kept it in `utils.py` because it makes more sense for `deployment_state.py` and `autoscaling_policy.py` to import from `utils.py` rather than each other.

Let me know what you think. Is there another common file that this method could be added to?

File: python/ray/serve/tests/test_target_capacity.py (hunk 13 above)
Comment (shrekris-anyscale) [reply python/ray/serve/tests/test_target_capacity.py]: Good idea, I caught the `ValueError` as suggested.

File: python/ray/serve/tests/test_target_capacity.py (hunk 10 above)
Comment (shrekris-anyscale) [reply python/ray/serve/tests/test_target_capacity.py:478]: The test now `deepcopies` the `config` before mutating it.

File: python/ray/serve/tests/test_target_capacity.py (hunk 9 above)
Comment (shrekris-anyscale) [reply python/ray/serve/tests/test_target_capacity.py]: Sounds good, I used a pydantic-typed `config`.

File: python/ray/serve/tests/test_target_capacity.py (hunk 14 above)
Comment (shrekris-anyscale) [reply python/ray/serve/tests/test_target_capacity.py:457]: Synced offline.

File: python/ray/serve/_private/application_state.py (hunk 17 above)
Comment (zcin) [reply python/ray/serve/_private/application_state.py:386]: I don't think this currently has any side effects. However in the long term instead of storing the app configs in the [controller](https://github.com/ray-project/ray/blob/master/python/ray/serve/_private/controller.py#L982-L988), I think it's probably better to fetch them from the application state manager. The target state is the place that we store the full config for an application, so would it be better if we don't clear the target config?

File: python/ray/serve/_private/deployment_state.py [hunk 22]
```diff
@@ -1912,6 +1933,39 @@ def _check_curr_status(self) -> Tuple[bool, bool]:
...
+        )
+
+        autoscaling_policy = self._target_state.info.autoscaling_policy
+        if autoscaling_policy is None:
+            has_reached_steady_state = (
+                self.curr_status_info.status_trigger
+                != DeploymentStatusTrigger.CONFIG_UPDATE_STARTED
+                or num_replicas_running_at_target_version
+                == self._target_state.target_num_replicas
+            )
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/deployment_state.py:1949]: This branch is not used (or tested). I added it for completeness' sake, but it may be better to remove it altogether. I'm not sure what to replace it with though.

File: python/ray/serve/_private/application_state.py (hunk 17 above)
Comment (shrekris-anyscale) [reply python/ray/serve/_private/application_state.py]: Makes sense. I updated the logic to store the config and clear the rest of the target state.

File: python/ray/serve/_private/deployment_state.py (hunk 22 above)
Comment (shrekris-anyscale) [reply python/ray/serve/_private/deployment_state.py]: I replaced the branch with an `assertion`.

File: python/ray/serve/tests/test_target_capacity.py (hunk 11 above)
Comment (shrekris-anyscale) [reply python/ray/serve/tests/test_target_capacity.py]: Good call, I added logic to check for specific values.

File: python/ray/serve/_private/deployment_state.py [hunk 23]
```diff
@@ -1522,10 +1528,17 @@ def deploy(self, deployment_info: DeploymentInfo) -> bool:
                 deployment_info.target_capacity,
             )
 
+        # Only allow the deployment status to become UPSCALING or DOWNSCALING
+        # if the deployment is not currently UPDATING.
+        allow_scaling_statuses = (
+            self.curr_status_info.status is not DeploymentStatus.UPDATING
+        )
+
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/deployment_state.py]: This is necessary if a deployment that's still `UPDATING` is redeployed before it has finished `UPDATING`. It should not enter `UPSCALING`/`DOWNSCALING` until it has finished `UPDATING`.

File: python/ray/serve/_private/deployment_state.py [hunk 24]
```diff
@@ -1595,55 +1578,70 @@ def autoscale(
...
             f"current_num_ongoing_requests: {current_num_ongoing_requests}, "
             f"current handle queued queries: {current_handle_queued_queries}."
         )
 
         new_info = copy(self._target_state.info)
-        new_info.set_autoscaled_num_replicas(clipped_decision_num_replicas)
         new_info.version = self._target_state.version.code_version
+
+        has_reached_steady_state = self._has_reached_steady_state()
+        if has_reached_steady_state:
```
Comment (edoakes) [inline python/ray/serve/_private/deployment_state.py:1602]: unclear what "steady_state" means, please leave a comment (or name it something more self-describing if possible!)

File: python/ray/serve/_private/deployment_state.py [hunk 25]
```diff
@@ -1595,55 +1578,70 @@ def autoscale(
...
+            decision_num_replicas,
+            status_trigger=new_status_trigger,
+            allow_scaling_statuses=has_reached_steady_state,
+        )
+
+    def _has_reached_steady_state(self) -> bool:
+        """Whether or not this deployment has reached steady-state.
+
+        This method should only be used for autoscaling deployments. It raises
+        an assertion error otherwise.
```
Comment (edoakes) [inline python/ray/serve/_private/deployment_state.py:1618]: again, don't know what "steady state" is -- this comment does not help me understand what the method is doing in any way

**Review by edoakes — APPROVED** [review]

**Comment by edoakes** [comment]
Relevant test failure: https://buildkite.com/ray-project/premerge/builds/13364#018c249e-e466-44ee-aab8-31dee44becc1/5202-5768

File: python/ray/serve/_private/deployment_state.py (hunk 24 above)
Comment (shrekris-anyscale) [reply python/ray/serve/_private/deployment_state.py]: I renamed the method  to `_has_reached_autoscaling_bounds()`.

File: python/ray/serve/_private/deployment_state.py (hunk 25 above)
Comment (shrekris-anyscale) [reply python/ray/serve/_private/deployment_state.py]: I renamed the method and changed the docstring. Let me know if it's clearer.

File: python/ray/serve/tests/test_target_capacity.py [hunk 26]
```diff
@@ -373,6 +381,458 @@ def test_autoscaling_scale_to_zero(
...
+            timeout=20,
+        )
+
+        # Clear requests and check that application scales down.
+        ray.get(request_signal.send.remote())
+        results = [request.result() for request in requests]
+        assert results == ["Hello world!"] * (4 * max_replicas)
+        wait_for_condition(
+            self.check_num_replicas,
+            expected_num_replicas=int(0.5 * initial_replicas),
```
Comment (zcin) [inline python/ray/serve/tests/test_target_capacity.py:723]: Curious (my understanding may be off): why is the number of replicas expected to scale down to 0.5 * 30 instead of 0.5 * 10 after requests have been sent? I'm confused as to why the replicas are limited to `[initial_replicas, max_replicas]` instead of `[min_replicas, max_replicas]` while target capacity is increasing and traffic is being sent. I thought `initial_replicas` was disregarded after the deployment starts autoscaling. Does the meaning of `initial_replicas` change with `target_capacity`?

File: python/ray/serve/tests/test_autoscaling_policy.py [hunk 27]
```diff
@@ -1389,6 +1392,260 @@ def check_num_replicas(num: int):
     assert existing_pid in pids
 
 
+@pytest.mark.skipif(sys.platform == "win32", reason="Failing on Windows.")
+def test_autoscaling_status_changes(serve_instance):
+    """Test status changes when autoscaling deployments are deployed."""
```
Comment (zcin) [inline python/ray/serve/tests/test_autoscaling_policy.py:1397]: This is a really long test - could you describe in the docstring what the test does in more detail? This could help people debug flaky/failing tests in the future. E.g: [example](https://github.com/ray-project/ray/blob/master/python/ray/serve/tests/unit/test_deployment_state.py#L1991-L2000)

File: python/ray/serve/_private/deployment_state.py [hunk 28]
```diff
@@ -1595,55 +1579,76 @@ def autoscale(
...
+        )
+        upper_bound = get_capacity_adjusted_num_replicas(
+            autoscaling_policy.config.max_replicas,
+            self._target_state.info.target_capacity,
+        )
+        has_reached_autoscaling_bounds = (
+            self.curr_status_info.status_trigger
+            != DeploymentStatusTrigger.CONFIG_UPDATE_STARTED
+            or lower_bound <= num_replicas_running_at_target_version <= upper_bound
+        )
```
Comment (zcin) [inline python/ray/serve/_private/deployment_state.py:1648]: This makes `_has_reached_autoscaling_bounds` return False if the status trigger is `CONFIG_UPDATE_STARTED`, but the name of the method implies that it's only checking for autoscaling bounds, which is a bit confusing to me. Can this check be moved outside of the method?

Also, can you add a comment explaining why we shouldn't allow scaling statuses if status trigger is config update started?

File: python/ray/serve/_private/deployment_state.py [hunk 29]
```diff
@@ -1595,55 +1579,76 @@ def autoscale(
...
 
         new_info = copy(self._target_state.info)
-        new_info.set_autoscaled_num_replicas(clipped_decision_num_replicas)
         new_info.version = self._target_state.version.code_version
+
+        has_reached_autoscaling_bounds = self._has_reached_autoscaling_bounds()
+        if has_reached_autoscaling_bounds:
+            new_status_trigger = DeploymentStatusTrigger.AUTOSCALING
+        else:
+            new_status_trigger = self.curr_status_info.status_trigger
```
Comment (zcin) [inline python/ray/serve/_private/deployment_state.py:1606]: Nit: this isn't strictly necessary since if `allow_scaling_statuses = False`, `set_target_state` won't touch the status.

**Review by zcin — APPROVED** [review]
Nice work! A few small nits

File: python/ray/serve/tests/test_autoscaling_policy.py (hunk 27 above)
Comment (shrekris-anyscale) [reply python/ray/serve/tests/test_autoscaling_policy.py:1397]: Thanks for the example! The test is still a WIP, but I added a docstring explaining what it does. Let me know what you think.

**Comment by shrekris-anyscale** [comment]
@edoakes @zcin All relevant tests are passing. If there's no more feedback, this change is ready to merge.
